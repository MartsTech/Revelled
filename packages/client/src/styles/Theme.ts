const Theme = {
  // neutrals (lightest to darkest)
  white: "#FFF",
  black: "#000",
};

export default Theme;
