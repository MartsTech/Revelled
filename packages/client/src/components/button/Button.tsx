import type {
  ButtonHTMLAttributes,
  DetailedHTMLProps,
  FC,
  ReactNode,
} from "react";
import styled from "styled-components";
import Theme from "styles/Theme";
import Accent from "./colors/Accent";
import Primary from "./colors/Primary";
import Secondary from "./colors/Secondary";
import Transparent from "./colors/Transparent";
import Big from "./sizes/Big";
import Small from "./sizes/Small";
import Tiny from "./sizes/Tiny";

type ButtonColor = "primary" | "secondary" | "accent" | "transparent";

type ButtonSize = "big" | "small" | "tiny";

type ButtonProps = DetailedHTMLProps<
  ButtonHTMLAttributes<HTMLButtonElement>,
  HTMLButtonElement
> & {
  size?: ButtonSize;
  color?: ButtonColor;
  loading?: boolean;
  icon?: ReactNode;
  theme?: typeof Theme;
};

const Button: FC<ButtonProps> = ({
  children,
  size = "big",
  color = "primary",
  disabled = false,
  loading = false,
  icon,
  ref,
  onClick,
  ...props
}) => {
  return (
    <StyledButton
      disabled={disabled || loading}
      color={color}
      size={size}
      data-testid="button"
      onClick={(e) => {
        e.preventDefault();
        typeof onClick === "function" && onClick(e);
      }}
      {...props}
    >
      <StyledContent>
        {icon && <StyledIconContainer>{icon}</StyledIconContainer>}
        {children}
      </StyledContent>
      {loading && (
        <StyledSpinnerContainer>
          {/* <Spinner size={size === "small" ? "2" : "4"} /> */}
        </StyledSpinnerContainer>
      )}
    </StyledButton>
  );
};

export default Button;

const StyledButton = styled.button<{
  color: ButtonColor;
  size: ButtonSize;
}>`
  display: flex;
  border: none;
  color: ${({ theme }) => theme.colors.button};
  font-size: ${({ theme }) => theme.fontSize["sm"]};
  font-weight: bold;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;

  transition-property: background-color, border-color, color, fill, stroke,
    opacity, box-shadow, transform, filter, backdrop-filter;
  transition-duration: 200ms;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);

  :focus {
    box-shadow: inset 0 0 0 2px currentColor;
  }

  :disabled {
    &&& {
      background-color: ${({ theme }) => theme.colors.primary[700]};
      cursor: not-allowed;
    }
  }

  ${({ color }) =>
    color === "primary"
      ? Primary
      : color === "secondary"
      ? Secondary
      : color === "accent"
      ? Accent
      : color === "transparent" && Transparent}

  ${({ size }) =>
    size === "big" ? Big : size === "small" ? Small : size === "tiny" && Tiny}
`;

const StyledContent = styled.span`
  display: flex;
  align-items: center;
`;

const StyledIconContainer = styled.span`
  margin-right: 0.5rem;
  align-items: center;
`;

const StyledSpinnerContainer = styled.span`
  position: absolute;
`;
