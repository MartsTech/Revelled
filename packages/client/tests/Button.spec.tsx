import Button from "components/button";
import { render, screen } from "@testing-library/react";
import Theme from "styles/Theme";

describe("Button", () => {
  it("should render a button with Click me text", () => {
    render(<Button theme={Theme}>Click me</Button>);
    const button = screen.getByTestId("button");

    expect(button).toBeInTheDocument();
  });

  it("should render a disabled button if loading prop is passed", () => {
    render(
      <Button loading={true} theme={Theme}>
        Click me
      </Button>
    );
    const button = screen.getByTestId("button");

    expect(button).toBeDisabled();
  });

  it("should render a disabled button if disabled prop is passed", () => {
    render(
      <Button disabled={true} theme={Theme}>
        Click me
      </Button>
    );
    const button = screen.getByTestId("button");

    expect(button).toBeDisabled();
  });

  it("should render an icon if icon props is passed", () => {
    render(
      <Button theme={Theme} icon={<div className="flex">Fake icon</div>}>
        Click me
      </Button>
    );
    const icon = screen.getByText("Fake icon");

    expect(icon).toBeVisible();
  });

  it("should match snapshot", () => {
    const { container } = render(
      <Button disabled={true} theme={Theme}>
        Click me
      </Button>
    );

    expect(container).toMatchSnapshot();
  });
});
