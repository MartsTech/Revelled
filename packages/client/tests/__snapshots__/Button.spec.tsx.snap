// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border: none;
  color: var(--color-button-text);
  font-size: 0.875rem;
  font-weight: bold;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-transition-property: background-color,border-color,color,fill,stroke, opacity,box-shadow,-webkit-transform,filter,backdrop-filter;
  -webkit-transition-property: background-color,border-color,color,fill,stroke, opacity,box-shadow,transform,filter,backdrop-filter;
  transition-property: background-color,border-color,color,fill,stroke, opacity,box-shadow,transform,filter,backdrop-filter;
  -webkit-transition-duration: 200ms;
  transition-duration: 200ms;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  background-color: var(--color-accent);
  padding: 0.625rem 2.5rem;
  border-radius: 0.5rem;
}

.c1:focus {
  box-shadow: inset 0 0 0 2px currentColor;
}

.c1:disabled.c0:disabled.c0:disabled {
  background-color: var(--color-primary-700);
  cursor: not-allowed;
}

.c1:hover {
  background-color: var(--color-accent-hover);
}

.c1:disabled {
  color: var(--color-accent-disabled);
  background-color: var(--color-accent-hover);
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
  <button
    class="c0 c1"
    color="primary"
    data-testid="button"
    disabled=""
  >
    <span
      class="c2"
    >
      Click me
    </span>
  </button>
</div>
`;
